`run` environment gets old setuptools from wrong Python

In https://github.com/coherent-oss/system/issues/17#issuecomment-2334917629, we've adopted `pipx` to run a cli tool (`coherent.test`). The [command is simple](https://github.com/coherent-oss/system/blob/518bcf3f4d0a826022de0a993f25cb9b84382952/.github/workflows/main.yml#L60):

```
pipx run --python 3.12 coherent.test
```

One of the dependencies of `coherent.test` is `setuptools` (by way of `setuptools-scm`).

Nevertheless, setuptools fails to import:

```
  File "/opt/pipx/.cache/d1eeadd94b582be/lib/python3.12/site-packages/setuptools_scm/_integration/setuptools.py", line 10, in <module>
    import setuptools
  File "/opt/pipx/shared/lib/python3.10/site-packages/setuptools/__init__.py", line 10, in <module>
    import distutils.core
ModuleNotFoundError: No module named 'distutils'
```

As you can see from the traceback, even though Python 3.12 was indicated, the dependency on `setuptools` was satisfied by the pipx install under Python 3.10. Under Python 3.10, one _might_ expect distutils to be supplied by the stdlib, so _might_ install Setuptools in such a way that it relies on the stdlib, but on Python 3.12, distutils is gone, so relies entirely on the vendored copy of distutils.

It seems a bug to me that pipx would allow any dependency to be satisfied by an environment for another Python version, especially if `--system-site-packages` wasn't part of the command.

This issue occurs on both Ubuntu and Windows in GitHub CI, so a fairly standard environment, but not one over which I have much control.

Any advice on why the wrong Setuptools is being involved here?

For posterity, here's the workflow:

```yaml
name: tests

on:
  push:

jobs:
  test:
    strategy:
      fail-fast: false
      matrix:
        python:
        - "3.12"
        platform:
        - ubuntu-latest
        - macos-latest
        - windows-latest

    runs-on: ${{ matrix.platform }}

    steps:
      - uses: actions/checkout@v4
      - name: Setup Python
        uses: actions/setup-python@v5
        with:
          python-version: ${{ matrix.python }}
      - name: Run tests for ${{ matrix.python }} (${{ runner.os }})
        run: pipx run --python ${{ matrix.python}} --spec setuptools python -c "import setuptools"
```

And here's the output:

```
Run pipx run --python 3.12 --spec setuptools python -c "import setuptools"
⚠️  python is already on your PATH and installed at
    /opt/hostedtoolcache/Python/3.12.5/x64/bin/python. Downloading and running
    anyway.
creating virtual environment...
installing setuptools...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/pipx/shared/lib/python3.10/site-packages/setuptools/__init__.py", line 10, in <module>
    import distutils.core
ModuleNotFoundError: No module named 'distutils'
Error: Process completed with exit code 1.
```

--- a/changelog.d/1539.bugfix.md
+++ b/changelog.d/1539.bugfix.md
@@ -1,0 +1,1 @@
+Remove `setuptools` from shared libs to prevent version conflicts when app venvs use a different Python.

--- a/src/pipx/shared_libs.py
+++ b/src/pipx/shared_libs.py
@@ -109,6 +109,15 @@ class _SharedLibs:
             pip_args.append("--force-reinstall")
             self.upgrade(pip_args=pip_args, verbose=verbose, raises=True)
 
+            # Remove setuptools from shared libs to prevent it from leaking into
+            # app venvs via the .pth file. On Python < 3.12, venv bundles setuptools
+            # via ensurepip, but a 3.10 setuptools breaks when imported under 3.12+
+            # because distutils was removed from the stdlib.
+            run_subprocess(
+                [self.python_path, "-m", "pip", "--no-input", "uninstall", "-y", "setuptools"],
+                capture_stderr=False,
+            )
+
     @property
     def is_valid(self) -> bool:
         if self.python_path.is_file():

--- a/tests/test_shared_libs.py
+++ b/tests/test_shared_libs.py
@@ -1,4 +1,6 @@
+import json
 import os
+import subprocess
 import time
 from pathlib import Path
 from unittest.mock import patch
@@ -65,6 +67,19 @@ def test_venv_python_is_valid_existing_interpreter(tmp_path: Path) -> None:
     pyvenv_cfg.write_text(f"home = {original_python_dir}\nversion = 3.12.0\n")
 
     assert shared_libs._venv_python_is_valid(python_exe) is True
+
+
+def test_shared_libs_excludes_setuptools(pipx_ultra_temp_env: None) -> None:
+    shared_libs.shared_libs.create(verbose=True, pip_args=[])
+    result = subprocess.run(
+        [str(shared_libs.shared_libs.python_path), "-m", "pip", "list", "--format=json"],
+        capture_output=True,
+        text=True,
+        check=True,
+    )
+    installed = {pkg["name"].lower() for pkg in json.loads(result.stdout)}
+    assert "pip" in installed
+    assert "setuptools" not in installed
 
 
 def test_venv_python_is_valid_non_windows() -> None:

